sdf3-v2.go

```diff
@@ -0,0 +1,676 @@
+	flags       int // offset to children is stored in the upper 30 bits, lower two bits are used as flags
+	left, right int64
+	// either:
+	// - left and right clipping plane values (reinterpret them as float64)
+	// - or offset intwo the mesh list and the number of elements that belong
+	//   to this leaf
+}
+
+func (b *bihInternal) is_leaf() bool {
+	return (b.flags & 3) == LEAF
+}
+
+func (b *bihInternal) is_x() bool {
+	return (b.flags & 3) == X_CLIP
+}
+
+func (b *bihInternal) is_y() bool {
+	return (b.flags & 3) == Y_CLIP
+}
+
+func (b *bihInternal) is_z() bool {
+	return (b.flags & 3) == Z_CLIP
+}
+
+func (b *bihInternal) leftClip() float64 {
+	return *(*float64)(unsafe.Pointer(&b.left))
+}
+func (b *bihInternal) rightClip() float64 {
+	return *(*float64)(unsafe.Pointer(&b.right))
+}
+
+func max(a, b float64) float64 {
+	if a > b {
+		return a
+	} else {
+		return b
+	}
+}
+
+func min(a, b float64) float64 {
+	if a < b {
+		return a
+	} else {
+		return b
+	}
+}
+
+func minDistBox(target r3.Vec, bb r3.Box) float64 {
+	// pretty cool formula based on this stackoverflow answer:
+	// https://math.stackexchange.com/questions/2133217/minimal-distance-to-a-cube-in-2d-and-3d-from-a-point-lying-outside
+	dx := max(0, max(target.X-bb.Max.X, bb.Min.X-target.X))
+	dy := max(0, max(target.Y-bb.Max.Y, bb.Min.Y-target.Y))
+	dz := max(0, max(target.Z-bb.Max.Z, bb.Min.Z-target.Z))
+	return dx*dx + dy*dy + dz*dz - 1e-6
+}
+
+type Mesh struct {
+	Edge_adj map[[2]int]int // edge to triangles mapping for adjacency
+	Indices  [][3]int       // indices into the vertex list
+	Vertices []r3.Vec
+	Bb       r3.Box
+}
+
+func NewMesh(model []r3.Triangle, vertTol float64) Mesh {
+	vertices := make([]r3.Vec, 0)
+	mesh := make([][3]int, len(model))
+	bb := r3.Box{
+		Min: r3.Vec{math.MaxFloat64, math.MaxFloat64, math.MaxFloat64},
+		Max: r3.Vec{-math.MaxFloat64, -math.MaxFloat64, -math.MaxFloat64},
+	}
+
+	vert_cache := make(map[[3]int64]int)
+	// copy the model into the vertex and mesh arrays
+	hvertTol := 0.5 * vertTol
+	for i, tri := range model {
+		for j, v := range tri {
+			// look for a vertex within tolerance
+			index := [3]int64{
+				int64((v.X + hvertTol) / vertTol),
+				int64((v.Y + hvertTol) / vertTol),
+				int64((v.Z + hvertTol) / vertTol),
+			}
+			if vidx, ok := vert_cache[index]; !ok {
+				vertices = append(vertices, v)
+				bb.Min = minVec(bb.Min, v)
+				bb.Max = maxVec(bb.Max, v)
+				mesh[i][j] = len(vertices) - 1
+				vert_cache[index] = len(vertices) - 1
+			} else {
+				mesh[i][j] = vidx
+			}
+		}
+	}
+
+	// generate edge adjacency list and calculate pseudonormals
+	edges := make(map[[2]int]int)
+	for i, tri := range mesh {
+		id := [2]int{tri[0], tri[1]}
+		edges[id] = i
+		id[0] = tri[1]
+		id[1] = tri[2]
+		edges[id] = i
+		id[0] = tri[2]
+		id[1] = tri[0]
+		edges[id] = i
+	}
+
+	return Mesh{
+		Edge_adj: edges,
+		Indices:  mesh,
+		Vertices: vertices,
+		Bb:       bb,
+	}
+}
+
+// TODO figure out what to make public
```

Comment:
For now its OK to leave stuff exported since it will be living in the experimental `sdfexp` package. That will give us time to figure out what works as an exported field and what should be private.

Reply:
Sounds good to me! For further reference the edge adjacency list, edge and vertex normals are all currently unused